youtube/innertube_caption.py

```diff
@@ -0,0 +1,92 @@
+    if resp:
+        caption_data = json.loads(resp)
+    else:
+        return None
+    vtt_body = caption_data['actions'][0]['updateEngagementPanelAction']['content']['transcriptRenderer']['content']['transcriptSearchPanelRenderer']['body']['transcriptSegmentListRenderer'].get('initialSegments')
+    if vtt_body:
+        return caption_data
+    else:
+        print('Unable to find vtt_body, retrying request with continuation params')
+        continuation_submenu = caption_data['actions'][0]['updateEngagementPanelAction']['content']['transcriptRenderer']['content']['transcriptSearchPanelRenderer']['footer']['transcriptFooterRenderer']['languageMenu']['sortFilterSubMenuRenderer']['subMenuItems']
+        for item in continuation_submenu:
+            if lang in item['title']:
+                params_new = item['continuation']['reloadContinuationData']['continuation']
+                payload['params'] = params_new
+                resp_new = util.fetch_url(caption_api, headers=header, data=json.dumps(payload), report_text=f'Fetching captions for {video_id}')
+                caption_data_new = json.loads(resp_new.decode())
+                vtt_body_new = caption_data_new['actions'][0]['updateEngagementPanelAction']['content']['transcriptRenderer']['content']['transcriptSearchPanelRenderer']['body']['transcriptSegmentListRenderer'].get('initialSegments')
+                if vtt_body_new:
+                    print('Got vtt_body from continuation request')
+                else:
+                    print('Still not getting vtt_body with continuation.\nReturning the new_caption_data as is.')
+                return caption_data_new
+
+def convert_milliseconds_to_hhmmss_optimized(ms):
+    if not isinstance(ms, int):
+        ms = int(ms)
+    total_seconds = ms // 1000
+    milliseconds = ms % 1000
+    hours = total_seconds // 3600
+    total_seconds %= 3600
+    minutes = total_seconds // 60
+    seconds = total_seconds % 60
+    return f"{hours:02}:{minutes:02}:{seconds:02}.{milliseconds:03}"
+
+def webvtt_from_caption_data(caption_data: dict = {}):
+    if not caption_data:
+        return None
+    vtt_body = caption_data['actions'][0]['updateEngagementPanelAction']['content']['transcriptRenderer']['content']['transcriptSearchPanelRenderer']['body']['transcriptSegmentListRenderer'].get('initialSegments')
+    if not vtt_body:
+        print('Unable to find vtt_body in the caption data')
+        return 'WEBVTT\n'
+
+    vtt_content = [ 'WEBVTT\n' ]
+
+    for item in vtt_body:
+        if item.get('transcriptSegmentRenderer'):
+            vtt_item = item['transcriptSegmentRenderer']
+            ms_to_vtt_time = convert_milliseconds_to_hhmmss_optimized
+            start_time = ms_to_vtt_time(vtt_item['startMs'])
+            end_time = ms_to_vtt_time(vtt_item['endMs'])
+            running_text = vtt_item['snippet']['runs'][0]['text']
+            webvtt_time_line = str.format("""{} --> {}\n{}\n""", start_time, end_time, running_text)
```

Comment:
I'm not sure why, but I think there's a mistake somewhere with how you're inserting newlines. For most videos, I'm seeing 3 - 4 lines of captions at a time at the start of the video, whereas when I view it on YouTube it correctly displays only 2 lines at a time.

Reply:
Ok, after reviewing the output of my vtt construction, it seems that the vtt part lacks newline at the end of running text.

Also the vtt is served as `text/html` instead of `text/vtt`.

Will get back to fixing this later.